File: frontend/src/types/stats.ts
/** Mirrors backend/app/schemas/stats.py (StatsResponse). Weeks are ISO date strings
 * (Monday of the week). See that file / stats_service.py for the scoping rules: the
 * time-series respect the date/instrument filters; cell/sample/credit figures are
 * current "now" snapshots. */

export interface WeekPoint {
  week: string;
  runs: number;
  samples: number;
}

export interface InstrumentThroughput {
  serial: string;
  name: string | null;
  runs: number;
  cell_uses: number;
}

export interface MovieHoursSlice {
  movie_hours: number;
  count: number;
}

export interface ThroughputStats {
  series: WeekPoint[];
  per_instrument: InstrumentThroughput[];
  movie_hours_mix: MovieHoursSlice[];
}

export interface DepthSlice {
  uses: number;
  cells: number;
}

export interface AvgUsesPoint {
  week: string;
  avg_uses: number;
}

export interface WellFillPoint {
  week: string;
  filled: number;
  capacity: number;
}

export interface WindowWaste {
  full_3_uses: number;
  expired_early: number;
}

export interface ReuseStats {
  depth_distribution: DepthSlice[];
  avg_uses_trend: AvgUsesPoint[];
  well_fill: WellFillPoint[];
  window_waste: WindowWaste;
}

export interface OutcomeSlice {
  status: string;
  count: number;
}

export interface FailureRatePoint {
  week: string;
  failed: number;
  total: number;
}

export interface CreditFunnel {
  needs_report: number;
  reported: number;
  awaiting: number;
  received: number;
}

export interface FailureStats {
  outcomes: OutcomeSlice[];
  failure_rate_trend: FailureRatePoint[];
  credit_funnel: CreditFunnel;
}

export interface StatusSlice {
  status: string;
  count: number;
}

export interface ImportVolumePoint {
  week: string;
  imported: number;
}

export interface InventoryStats {
  cell_status: StatusSlice[];
  sample_funnel: StatusSlice[];
  import_volume: ImportVolumePoint[];
}

export interface HeadlineStats {
  runs_completed: number;
  samples_completed: number;
  avg_uses_per_cell: number;
  pct_reaching_use3: number;
  failure_rate: number;
  well_fill_pct: number;
  cells_awaiting_credit: number;
  credits_received: number;
}

export interface StatsResponse {
  headline: HeadlineStats;
  throughput: ThroughputStats;
  reuse: ReuseStats;
  failures: FailureStats;
  inventory: InventoryStats;
}

File: backend/app/schemas/stats.py
"""Response shape for GET /api/stats - the aggregated figures behind the Stats page.

Grouped to mirror the four dashboard sections (throughput, reuse, failures, inventory)
plus a flat headline block for the KPI tiles. All aggregation lives in
services/stats_service.py; nothing here computes anything.

Note on scoping (kept consistent with the Help copy): the time-series and throughput
figures respect the date-range/instrument filters; the "now" snapshots - cell status,
sample funnel, and the PacBio credit funnel - describe current outstanding state and are
not date-filtered (a credit still owed doesn't stop mattering because it fell outside the
window)."""
from datetime import date

from pydantic import BaseModel


# --- Throughput & run rate ---
class WeekPoint(BaseModel):
    week: date  # Monday of the ISO week the figures fall in
    runs: int
    samples: int


class InstrumentThroughput(BaseModel):
    serial: str
    name: str | None
    runs: int
    cell_uses: int


class MovieHoursSlice(BaseModel):
    movie_hours: int
    count: int


class ThroughputStats(BaseModel):
    series: list[WeekPoint]
    per_instrument: list[InstrumentThroughput]
    movie_hours_mix: list[MovieHoursSlice]


# --- Reuse & utilisation ---
class DepthSlice(BaseModel):
    uses: int  # 1, 2 or 3
    cells: int


class AvgUsesPoint(BaseModel):
    week: date
    avg_uses: float


class WellFillPoint(BaseModel):
    week: date
    filled: int
    capacity: int  # runs that week * 8 wells


class WindowWaste(BaseModel):
    full_3_uses: int  # terminal cells that got all 3 uses out of the 108h window
    expired_early: int  # terminal cells whose window lapsed before all 3 uses were spent


class ReuseStats(BaseModel):
    depth_distribution: list[DepthSlice]
    avg_uses_trend: list[AvgUsesPoint]
    well_fill: list[WellFillPoint]
    window_waste: WindowWaste


# --- Failures & credits ---
class OutcomeSlice(BaseModel):
    status: str  # completed | failed | aborted
    count: int


class FailureRatePoint(BaseModel):
    week: date
    failed: int
    total: int  # completed + failed + aborted (uses that got a real verdict)


class CreditFunnel(BaseModel):
    needs_report: int  # failed/stopped cell, not yet raised with PacBio
    reported: int  # ever raised with PacBio
    awaiting: int  # raised, credit not yet received
    received: int  # credit landed


class FailureStats(BaseModel):
    outcomes: list[OutcomeSlice]
    failure_rate_trend: list[FailureRatePoint]
    credit_funnel: CreditFunnel


# --- Inventory & backlog ---
class StatusSlice(BaseModel):
    status: str
    count: int


class ImportVolumePoint(BaseModel):
    week: date
    imported: int


class InventoryStats(BaseModel):
    cell_status: list[StatusSlice]
    sample_funnel: list[StatusSlice]
    import_volume: list[ImportVolumePoint]


# --- Headline KPI tiles ---
class HeadlineStats(BaseModel):
    runs_completed: int
    samples_completed: int
    avg_uses_per_cell: float
    pct_reaching_use3: float  # 0-100
    failure_rate: float  # 0-100
    well_fill_pct: float  # 0-100
    cells_awaiting_credit: int
    credits_received: int


class StatsResponse(BaseModel):
    headline: HeadlineStats
    throughput: ThroughputStats
    reuse: ReuseStats
    failures: FailureStats
    inventory: InventoryStats
